src/containerapp/azext_containerapp/_clients.py

```diff
@@ -0,0 +1,723 @@
+from azure.cli.core.commands.client_factory import get_subscription_id
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+
+API_VERSION = "2021-03-01"
+NEW_API_VERSION = "2022-01-01-preview"
+POLLING_TIMEOUT = 60  # how many seconds before exiting
+POLLING_SECONDS = 2  # how many seconds between requests
+
+
+class PollingAnimation():
+    def __init__(self):
+        self.tickers = ["/", "|", "\\", "-", "/", "|", "\\", "-"]
+        self.currTicker = 0
+
+    def tick(self):
+        sys.stdout.write('\r')
+        sys.stdout.write(self.tickers[self.currTicker] + " Running ..")
+        sys.stdout.flush()
+        self.currTicker += 1
+        self.currTicker = self.currTicker % len(self.tickers)
+
+    def flush(self):
+        sys.stdout.flush()
+        sys.stdout.write('\r')
+        sys.stdout.write("\033[K")
+
+
+def poll(cmd, request_url, poll_if_status):  # pylint: disable=inconsistent-return-statements
+    try:
+        start = time.time()
+        end = time.time() + POLLING_TIMEOUT
+        animation = PollingAnimation()
+
+        animation.tick()
+        r = send_raw_request(cmd.cli_ctx, "GET", request_url)
+
+        while r.status_code in [200, 201] and start < end:
+            time.sleep(POLLING_SECONDS)
+            animation.tick()
+
+            r = send_raw_request(cmd.cli_ctx, "GET", request_url)
+            r2 = r.json()
+
+            if "properties" not in r2 or "provisioningState" not in r2["properties"] or not r2["properties"]["provisioningState"].lower() == poll_if_status:
+                break
+            start = time.time()
+
+        animation.flush()
+        return r.json()
+    except Exception as e:  # pylint: disable=broad-except
+        animation.flush()
+
+        delete_statuses = ["scheduledfordelete", "cancelled"]
+
+        if poll_if_status not in delete_statuses:  # Catch "not found" errors if polling for delete
+            raise e
+
+
+class ContainerAppClient():
+    @classmethod
+    def create_or_update(cls, cmd, resource_group_name, name, container_app_envelope, no_wait=False):
+        management_hostname = cmd.cli_ctx.cloud.endpoints.resource_manager
+        api_version = NEW_API_VERSION
+        sub_id = get_subscription_id(cmd.cli_ctx)
+        url_fmt = "{}/subscriptions/{}/resourceGroups/{}/providers/Microsoft.App/containerApps/{}?api-version={}"
+        request_url = url_fmt.format(
+            management_hostname.strip('/'),
+            sub_id,
+            resource_group_name,
+            name,
+            api_version)
+
+        r = send_raw_request(cmd.cli_ctx, "PUT", request_url, body=json.dumps(container_app_envelope))
+
+        if no_wait:
+            return r.json()
+        elif r.status_code == 201:
+            url_fmt = "{}/subscriptions/{}/resourceGroups/{}/providers/Microsoft.App/containerApps/{}?api-version={}"
+            request_url = url_fmt.format(
+                management_hostname.strip('/'),
+                sub_id,
```

Comment:
Are there any methods to call these REST in the Python SDK? Why do we need to implement the logic of requesting REST directly?

Reply:
OK